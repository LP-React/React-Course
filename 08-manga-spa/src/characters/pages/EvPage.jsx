import { CharacterList } from "../components/CharacterList"

export const EvPage = () => {
  return (
    <>
    <h1>Neon Genesis Evangelion</h1>
    <CharacterList manga={"Neon Genesis Evangelion"} />
    </>
  )
}
