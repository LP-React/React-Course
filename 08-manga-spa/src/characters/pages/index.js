export * from './DdPage'
export * from './EvPage'
export * from './OpPage'
export * from './SearchPage'
export * from './CharacterPage'