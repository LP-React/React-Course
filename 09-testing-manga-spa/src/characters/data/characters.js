export const characters = [
    {
        id: 'dd-koyama',
        name: 'Koyama',
        publisher: 'Inio Asano',
        manga: "Dead Dead Demon's Dededede Destruction"
    },
    {
        id: 'dd-kurihara',
        name: 'Kurihara',
        publisher: 'Inio Asano',
        manga: "Dead Dead Demon's Dededede Destruction"
    },
    {
        id: 'dd-nakagawa',
        name: 'Nakagawa',
        publisher: 'Inio Asano',
        manga: "Dead Dead Demon's Dededede Destruction"
    },
    {
        id: 'dd-ooba',
        name: 'Ooba',
        publisher: 'Inio Asano',
        manga: "Dead Dead Demon's Dededede Destruction"
    },
    {
        id: 'dd-tainuma',
        name: 'Tainuma',
        publisher: 'Inio Asano',
        manga: "Dead Dead Demon's Dededede Destruction"
    },
    {
        id: 'ev-asuka',
        name: 'Asuka',
        publisher: 'Hideaki Anno',
        manga: "Neon Genesis Evangelion"
    },
    {
        id: 'ev-kaworu',
        name: 'Kaworu',
        publisher: 'Hideaki Anno',
        manga: "Neon Genesis Evangelion"
    },
    {
        id: 'ev-misato',
        name: 'Misato',
        publisher: 'Hideaki Anno',
        manga: "Neon Genesis Evangelion"
    },
    {
        id: 'ev-rei',
        name: 'Rei',
        publisher: 'Hideaki Anno',
        manga: "Neon Genesis Evangelion"
    },
    {
        id: 'ev-shinji',
        name: 'Shinji',
        publisher: 'Hideaki Anno',
        manga: "Neon Genesis Evangelion"
    },
    {
        id: 'op-aiko',
        name: 'Aiko',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    },
    {
        id: 'op-kanie',
        name: 'Kanie',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    },
    {
        id: 'op-kou',
        name: 'Kou',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    },
    {
        id: 'op-masumi',
        name: 'Masumi',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    },
    {
        id: 'op-midori',
        name: 'Midori',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    },
    {
        id: 'op-punpun',
        name: 'Punpun',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    },
    {
        id: 'op-sachi',
        name: 'Sachi',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    },
    {
        id: 'op-takao',
        name: 'Takao',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    },
    {
        id: 'op-yuuichi',
        name: 'Yuuichi',
        publisher: 'Inio Asano',
        manga: "Oyasumi Punpun"
    }
];
