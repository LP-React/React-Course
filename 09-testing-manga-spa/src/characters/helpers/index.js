export * from './getCharacterByPublisher'
export * from './getCharacterById'
export * from './getCharacterByName'