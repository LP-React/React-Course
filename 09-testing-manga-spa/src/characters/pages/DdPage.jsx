import { CharacterList } from "../components/CharacterList"

export const DdPage = () => {
  return (
    <>
      <h1>Dead Dead Demon's Dededede Destruction</h1>
      <CharacterList manga="Dead Dead Demon's Dededede Destruction" />
    </>
  )
}
