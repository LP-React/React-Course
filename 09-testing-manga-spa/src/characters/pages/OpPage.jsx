import { CharacterList } from "../components/CharacterList"

export const OpPage = () => {
  return (
    <>
      <h1>Oyasumi PunPun</h1>

      <CharacterList manga={"Oyasumi Punpun"} />
    </>
  )
}
