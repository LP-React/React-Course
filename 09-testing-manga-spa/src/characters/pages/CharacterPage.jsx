import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getCharacterById } from "../helpers"
import { useMemo } from "react"

export const CharacterPage = () => {

    const { characterId } = useParams()

    const character = useMemo(() => getCharacterById(characterId), [characterId])

    const navigate = useNavigate()

    const onNavigateBack = () => {
        navigate(-1)
    }


    if (!character) {
        return <Navigate to={'/'} />
    }

    return (
        <div className="row mt-5">
            <div className="col-4">
                <img src={`/assets/characters/${character.id}.webp`} alt={character.name} className="img-thumbnail" />

            </div>
            <div className="col-3">
                <h3>{character.name}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item">From: {character.manga}</li>
                    <li className="list-group-item">Author: {character.publisher}</li>
                </ul>
                <button className="btn btn-outline-primary" onClick={onNavigateBack}>
                    Back
                </button>
            </div>
        </div>
    )
}
