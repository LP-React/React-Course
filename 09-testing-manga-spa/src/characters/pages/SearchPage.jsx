import { useLocation, useNavigate } from "react-router-dom"

import queryString from "query-string"

import { useForm } from "../../hooks/useForm"
import { CharacterCard } from "../components"
import { getCharacterByName } from "../helpers"

export const SearchPage = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const characters = getCharacterByName(q);


    const { searchText, onInputChange, onClear } = useForm({
        searchText: q
    });

    const onSearchSubmit = (event) => {
        event.preventDefault();
        navigate(`?q=${searchText}`)
    }

    return (
        <>
            <h1>Search</h1>

            <div className="row">

                <div className="col-5">
                    <h4>Searching...</h4>
                    <hr />

                    <form onSubmit={onSearchSubmit} aria-label="form">
                        <input type="text" placeholder="Search a character" name="searchText" autoComplete="off" className="form-control" value={searchText} onChange={onInputChange} />
                        <button className="btn btn-outline-primary mt-1">
                            Search
                        </button>
                    </form>

                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '')
                            ? <div className="alert alert-primary">Search a character</div>
                            : (characters.length === 0) && <div aria-label="alert" className="alert alert-danger">No character with <b>{q}</b></div>
                    }


                    {
                        characters.map(
                            character =>
                                <CharacterCard character={character} key={character.id} />
                        )
                    }
                </div>
            </div>
        </>
    )
}
