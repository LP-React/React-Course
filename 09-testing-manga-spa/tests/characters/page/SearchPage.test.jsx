import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { SearchPage } from "../../../src/characters/pages"

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate
}))

describe('Pruebas en <CharacterPage/>', () => {

        beforeEach(() => jest.clearAllMocks());

    test('Debe de mostrarse correctamente con valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot()
    })

    test('Debe de mostrar a Rei y el input con el valor del queryString', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=rei']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');

        expect(input.value).toBe('rei')

        const img = screen.getByRole('img');

        expect(img.src).toContain('/assets/characters/ev-rei.webp')

    })

    test('Debe de mostrar error cuando no se encuentra el personaje ("Laysson")', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=Laysson']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');

        expect(input.value).toBe('Laysson')

        const errorDiv = screen.getByLabelText('alert')

        expect(errorDiv).toBeTruthy()

        screen.debug()
    })



    test('Debe de llamar al Navigate al enviar el formulario', () => {

        const inputValue = 'Asuka'

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');

        fireEvent.change(input, {target: {
            name: 'searchText',
            value: inputValue
        }})

        expect(input.value).toBe(inputValue)

        const form = screen.getByRole('form')
        fireEvent.submit(form)
        
        expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`)
    })
})