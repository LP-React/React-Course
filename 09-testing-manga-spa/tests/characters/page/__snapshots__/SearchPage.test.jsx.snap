// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <CharacterPage/> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching...
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a character"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary"
      >
        Search a character
      </div>
    </div>
  </div>
</div>
`;
